lib/Eisbach_Methods.thy

```diff
@@ -574,4 +575,156 @@ method if_concl for pattern::bool methods m =
 method case_concl for pattern::bool methods m =
   has_concl pattern, m
 
+text \<open>@{text "all"} and @{text ";"} are dangerous for weakest-precondition reasoning, as the wp
+      goals usually need to be solved in the order they were generated. The @{text fwd_all}
+      and @{text fwd_all_new} methods apply on goals in order, taking into account goals solved
+      and added.\<close>
+
+ML \<open>
+
+fun restrict_goal i n tac =
+  (PRIMITIVE (Goal.restrict i n)) THEN
+  tac THEN
+  (PRIMITIVE (Goal.unrestrict i));
+
+local
+
+(* apply tac to all goals from i going forwards, taking into account any goals solved or generated *)
+fun tac_in_order (tac : int -> tactic) i thm =
+  if i > Thm.nprems_of thm orelse i < 1
+  then Seq.single thm
+  else
+    let
+      val st = restrict_goal i 1 (tac 1) thm (* only allow tac1 run on goal i *)
+    in
+      (* if the goal was solved, we want to stay on the same goal,
+         if the goal was transformed, we want to move to the next one, skipping any new
+         goals that were generated *)
+      Seq.maps (fn thm' => tac_in_order tac (i + 1 + Thm.nprems_of thm' - Thm.nprems_of thm) thm')
+               st
+    end
+
+in
+
+fun ALLGOALS_FWD tac = tac_in_order tac 1;
+
+end
+
+infix 1 FWD_ALL_NEW;
+
+(* this is like THEN_ALL_NEW but expressed with goal restriction and ALLGOALS_REV *)
+fun ((tac1 : int -> tactic) FWD_ALL_NEW (tac2 : int -> tactic)) (i : int) thm =
+  Seq.maps (fn thm' =>
+              if Thm.nprems_of thm' < Thm.nprems_of thm
+              then Seq.single thm'
+              else restrict_goal i (i + Thm.nprems_of thm' - Thm.nprems_of thm) (ALLGOALS_FWD tac2) thm'
+           ) (tac1 1 thm)
+
+\<close>
+
+method_setup fwd_all =
+  \<open>Method.text_closure >> (fn m => fn ctxt => fn facts =>
+   let
+     fun tac i st' =
+       Goal.restrict i 1 st'
+       |> method_evaluate m ctxt facts
+       |> Seq.map (Goal.unrestrict i)
+
```

Comment:
Tiny nitpick, but I think I would prefer this without the empty line.

Reply:
Done